tests/unit/test_endpoints_v2.py

```diff
@@ -0,0 +1,634 @@
+        ("amazon.com", False, False),
+        ("amazon.com", True, True),
+        ("foo", False, True),
+        ("foo", True, True),
+    ],
+)
+def test_is_valid_host_label(
+    standard_library, label, allow_subdomains, expected_output
+):
+    assert (
+        standard_library._is_valid_host_label(label, allow_subdomains)
+        == expected_output
+    )
+
+
+def test_partition_none_return_none(standard_library):
+    assert standard_library._aws_partition(None) is None
+
+
+@pytest.mark.parametrize("input_arn", [None, "arn:aws:this-is-not-an-arn:foo"])
+def test_parse_arn_none_return_none(standard_library, input_arn):
+    assert standard_library._aws_parse_arn(input_arn) is None
+
+
+@pytest.mark.parametrize(
+    "uri,expected_output",
+    [
+        (None, None),
+        (
+            ":/?#[]@!$&'()*+,;=%.",
+            "%3A%2F%3F%23%5B%5D%40%21%24%26%27%28%29%2A%2B%2C%3B%3D%25.",
+        ),
+        ("?foo=bar&baz=qux", "%3Ffoo%3Dbar%26baz%3Dqux"),
+    ],
+)
+def test_uri_encode(standard_library, uri, expected_output):
+    assert standard_library._uri_encode(uri) == expected_output
+
+
+def test_parse_url_none_return_none(standard_library):
+    assert standard_library._parse_url(None) is None
+
+
+def test_string_equals_wrong_type_raises(standard_library):
+    with pytest.raises(RuleSetError) as exc_info:
+        standard_library._string_equals(1, 2)
+    assert str(exc_info.value) == (
+        "Something went wrong during endpoint resolution: Both values must be strings.\n"
+        "value1: 1\n"
+        "value2: 2"
+    )
+
+
+def test_boolean_equals_wrong_type_raises(standard_library):
+    with pytest.raises(RuleSetError) as exc_info:
+        standard_library._boolean_equals(1, 2)
+    assert str(exc_info.value) == (
+        "Something went wrong during endpoint resolution: Both arguments must be booleans."
+        "\nvalue1: 1"
+        "\nvalue2: 2"
+    )
+
+
+def test_substring_wrong_type_raises(standard_library):
+    with pytest.raises(RuleSetError) as exc_info:
+        standard_library._substring(["h", "e", "l", "l", "o"], 0, 5, False)
+    assert str(exc_info.value) == (
+        "Something went wrong during endpoint resolution: Input must be a string.\n"
+        "string_input: ['h', 'e', 'l', 'l', 'o']"
+    )
+
+
+@pytest.mark.parametrize(
+    "input_string,start,stop,reverse,expected_output",
+    [
+        ("hello", 0, 5, False, "hello"),
+        ("hello", 0, -1, False, None),
+        ("hello", -5, -1, False, "hell"),
+        ("hello", -5, 5, False, "hello"),
+        ("hello", 0, 10, False, None),
+        ("hello", 3, 4, False, "l"),
+        ("hello", 0, 5, True, "hello"),
+        ("hello", 0, -1, True, None),
+        ("hello", -5, -1, True, ""),
+        ("hello", -5, 5, True, "hello"),
+        ("hello", 0, 10, True, None),
+        ("hello", 3, 4, True, "e"),
+        ("hello", 0, 3, True, "llo"),
+        ("hello", 3, 5, True, "he"),
```

Comment:
A number of these seem like redundant test cases. Are they all needed?

Reply:
it looks like none of them are needed anymore as the entire function is covered in the generated test cases. i've deleted.